internal/pkg/manager/workloadannotator/workloadannotator.go

```diff
@@ -0,0 +1,197 @@
+				return errors.Wrap(err, "retrieving profile")
+			}
+		}
+
+		return errors.Wrap(updateErr, "updating profile")
+	}, util.IsNotFoundOrConflict); err != nil {
+		return errors.Wrap(err, "updating SeccompProfile status")
+	}
+	hasActivePodsFinalizerString := "in-use-by-active-pods"
+	if len(linkedPods.Items) > 0 {
+		if err := util.Retry(func() error {
+			return util.AddFinalizer(ctx, r.client, sp, hasActivePodsFinalizerString)
+		}, util.IsNotFoundOrConflict); err != nil {
+			return errors.Wrap(err, "adding finalizer")
+		}
+	} else {
+		if err := util.Retry(func() error {
+			return util.RemoveFinalizer(ctx, r.client, sp, hasActivePodsFinalizerString)
+		}, util.IsNotFoundOrConflict); err != nil {
+			return errors.Wrap(err, "removing finalizer")
+		}
+	}
+	return nil
+}
+
+// getSeccompProfilesFromPod returns a slice of strings representing seccomp profiles required by the pod.
+// It looks first at the pod spec level, then in each container and init container, then in the annotations.
+func getSeccompProfilesFromPod(pod *corev1.Pod) []string {
+	profiles := []string{}
+	// try to get profile from pod securityContext
+	sc := pod.Spec.SecurityContext
+	if sc != nil && isOperatorSeccompProfile(sc.SeccompProfile) {
+		profiles = append(profiles, *sc.SeccompProfile.LocalhostProfile)
+	}
+	// try to get profile(s) from securityContext in pods
+	containers := append(pod.Spec.Containers, pod.Spec.InitContainers...)
+	contains := func(a []string, b string) bool {
+		for _, s := range a {
+			if s == b {
+				return true
+			}
+		}
+		return false
+	}
+	for i := range containers {
+		sc := containers[i].SecurityContext
+		if sc != nil && isOperatorSeccompProfile(sc.SeccompProfile) {
+			profileString := *containers[i].SecurityContext.SeccompProfile.LocalhostProfile
+			if !contains(profiles, profileString) {
+				profiles = append(profiles, profileString)
+			}
+		}
+	}
+	// try to get profile from annotations
+	annotation, hasAnnotation := pod.GetAnnotations()[corev1.SeccompPodAnnotationKey]
+	if hasAnnotation && strings.HasPrefix(annotation, "localhost/") {
+		profileString := strings.TrimPrefix(annotation, "localhost/")
+		if !contains(profiles, profileString) {
+			profiles = append(profiles, profileString)
+		}
+	}
+	return profiles
+}
+
+// isOperatorSeccompProfile checks whether a corev1.SeccompProfile object belongs to the operator.
+// SeccompProfiles controlled by the operator are of type "Localhost" and have a path of the form
+// "operator/namespace/profile-name.json".
+func isOperatorSeccompProfile(sp *corev1.SeccompProfile) bool {
+	if sp == nil || sp.Type != "Localhost" {
+		return false
+	}
+	if !strings.HasPrefix(*sp.LocalhostProfile, "operator/") {
+		return false
+	}
+	if !strings.HasSuffix(*sp.LocalhostProfile, ".json") {
+		return false
+	}
+	const pathParts = 3
+	return len(strings.Split(*sp.LocalhostProfile, "/")) == pathParts
+}
```

Comment:
Hm, adding a new bool field to operator-tracked profiles would be an alternative.

Reply:
The main issue was that by having the reconciler that received events for pods in the DaemonSet, each instance might receive the same event for the same pod, so it would try to update the same Seccomp profile; resulting in races. By moving that reconciler to the Deployment, that ensures that only the leader does this, and so it eliminates the race conditions on that side.